File: ios/Sources/SDUICore/JSONValue.swift
import Foundation

/// A type-safe representation of any JSON value.
///
/// The SDUI contract is intentionally open at the leaves: request bodies, list
/// item payloads, screen state and `custom.*` component props can hold arbitrary
/// JSON. `JSONValue` lets us carry that through the pipeline without losing type
/// information and without resorting to `Any`, so the whole model stays `Sendable`
/// and `Equatable`.
public enum JSONValue: Codable, Equatable, Hashable, Sendable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let b = try? container.decode(Bool.self) {
            self = .bool(b)
        } else if let d = try? container.decode(Double.self) {
            self = .number(d)
        } else if let s = try? container.decode(String.self) {
            self = .string(s)
        } else if let a = try? container.decode([JSONValue].self) {
            self = .array(a)
        } else if let o = try? container.decode([String: JSONValue].self) {
            self = .object(o)
        } else {
            throw DecodingError.dataCorruptedError(
                in: container, debugDescription: "Unsupported JSON value")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let s): try c.encode(s)
        case .number(let n): try c.encode(n)
        case .bool(let b):   try c.encode(b)
        case .object(let o): try c.encode(o)
        case .array(let a):  try c.encode(a)
        case .null:          try c.encodeNil()
        }
    }
}

// MARK: - Ergonomic access

public extension JSONValue {
    /// Subscript into an object member. Returns `nil` for non-objects or missing keys.
    subscript(_ key: String) -> JSONValue? {
        if case .object(let o) = self { return o[key] }
        return nil
    }

    /// Subscript into an array index. Returns `nil` when out of range or not an array.
    subscript(_ index: Int) -> JSONValue? {
        if case .array(let a) = self, a.indices.contains(index) { return a[index] }
        return nil
    }

    var stringValue: String? {
        switch self {
        case .string(let s): return s
        case .number(let n): return n == n.rounded() ? String(Int(n)) : String(n)
        case .bool(let b): return b ? "true" : "false"
        default: return nil
        }
    }

    var doubleValue: Double? {
        switch self {
        case .number(let n): return n
        case .string(let s): return Double(s)
        case .bool(let b): return b ? 1 : 0
        default: return nil
        }
    }

    var boolValue: Bool? {
        switch self {
        case .bool(let b): return b
        case .number(let n): return n != 0
        case .string(let s): return ["true", "1", "yes"].contains(s.lowercased())
        default: return nil
        }
    }

    var arrayValue: [JSONValue]? {
        if case .array(let a) = self { return a }
        return nil
    }

    var isNull: Bool {
        if case .null = self { return true }
        return false
    }

    /// Round-trips this value into any `Decodable` type. Used to lazily decode
    /// type-specific component props and modifiers without a giant enum.
    func decode<T: Decodable>(_ type: T.Type) -> T? {
        guard let data = try? JSONEncoder().encode(self) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }

    /// Navigates a dotted key path, e.g. "product.title" or "items.0.id".
    func value(at path: [String]) -> JSONValue? {
        var current: JSONValue? = self
        for segment in path {
            guard let node = current else { return nil }
            if let index = Int(segment) {
                current = node[index]
            } else {
                current = node[segment]
            }
        }
        return current
    }
}

File: android/sdui/src/main/java/dev/sdui/core/JsonValue.kt
package dev.sdui.core

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.doubleOrNull

/**
 * A type-safe representation of any JSON value — the Kotlin counterpart of the
 * iOS `JSONValue`.
 *
 * The SDUI contract is intentionally open at the leaves: request bodies, list
 * item payloads, screen state and `custom.*` component props can hold arbitrary
 * JSON. Modelling that as a sealed hierarchy (rather than `Any?`) lets us carry
 * the value through the whole pipeline without losing type information, keeps the
 * model comparable via [equals], and gives us the ergonomic accessors below so
 * the binding engine never has to reach for casts.
 */
@Serializable(with = JsonValueSerializer::class)
sealed class JsonValue {
    /** A JSON string leaf. */
    data class Str(val value: String) : JsonValue()

    /** A JSON number leaf, always stored as [Double] to match the contract. */
    data class Num(val value: Double) : JsonValue()

    /** A JSON boolean leaf. */
    data class Bool(val value: Boolean) : JsonValue()

    /** A JSON object; key order is preserved for deterministic rendering. */
    data class Obj(val value: Map<String, JsonValue>) : JsonValue()

    /** A JSON array. */
    data class Arr(val value: List<JsonValue>) : JsonValue()

    /** JSON `null`. Modelled explicitly so "present but null" differs from absent. */
    data object Null : JsonValue()

    // MARK: - Ergonomic access

    /**
     * Object-member access. Returns `null` for non-objects or missing keys so
     * callers can chain safely — this is the Kotlin analogue of the Swift
     * `subscript(key:)`.
     */
    operator fun get(key: String): JsonValue? = (this as? Obj)?.value?.get(key)

    /**
     * Array-index access. Returns `null` when out of range or not an array,
     * mirroring the Swift `subscript(index:)`.
     */
    operator fun get(index: Int): JsonValue? =
        (this as? Arr)?.value?.getOrNull(index)

    /**
     * The value coerced to a display string, or `null` when it cannot be. Whole
     * numbers render without a trailing `.0` so ids and counts read cleanly.
     */
    val stringValue: String?
        get() = when (this) {
            is Str -> value
            is Num -> if (value == kotlin.math.floor(value) && !value.isInfinite())
                value.toLong().toString() else value.toString()
            is Bool -> if (value) "true" else "false"
            else -> null
        }

    /** The value coerced to a [Double], or `null`. Strings and bools are parsed leniently. */
    val doubleValue: Double?
        get() = when (this) {
            is Num -> value
            is Str -> value.toDoubleOrNull()
            is Bool -> if (value) 1.0 else 0.0
            else -> null
        }

    /** The value coerced to a [Boolean], or `null`. Accepts `1/yes/true` in strings. */
    val boolValue: Boolean?
        get() = when (this) {
            is Bool -> value
            is Num -> value != 0.0
            is Str -> value.lowercase() in setOf("true", "1", "yes")
            else -> null
        }

    /** The underlying list when this is an array, otherwise `null`. */
    val arrayValue: List<JsonValue>?
        get() = (this as? Arr)?.value

    /** True when this is the explicit [Null] leaf. */
    val isNull: Boolean
        get() = this is Null

    /**
     * Navigates a dotted key path, e.g. `["product", "title"]` or
     * `["items", "0", "id"]`. Numeric segments index into arrays. Returns `null`
     * as soon as any segment is missing — a missing binding is never an error.
     */
    fun valueAt(path: List<String>): JsonValue? {
        var current: JsonValue? = this
        for (segment in path) {
            val node = current ?: return null
            current = segment.toIntOrNull()?.let { node[it] } ?: node[segment]
        }
        return current
    }

    /**
     * Decodes this value into any [@Serializable] type by round-tripping through
     * a [JsonElement]. Used to lazily decode type-specific component props and
     * modifiers without a closed enum of every component kind — the direct
     * analogue of the Swift `decode(_:)`.
     */
    inline fun <reified T> decode(): T? =
        runCatching { CODEC.decodeFromJsonElement<T>(toJsonElement()) }.getOrNull()

    /** Converts this value back into a kotlinx [JsonElement] for encoding/decoding. */
    fun toJsonElement(): JsonElement = when (this) {
        is Str -> JsonPrimitive(value)
        is Num -> JsonPrimitive(value)
        is Bool -> JsonPrimitive(value)
        is Obj -> JsonObject(value.mapValues { it.value.toJsonElement() })
        is Arr -> JsonArray(value.map { it.toJsonElement() })
        Null -> JsonNull
    }

    companion object {
        /**
         * A lenient JSON codec shared by [decode] and the parser. `ignoreUnknownKeys`
         * lets payloads carry fields a given renderer version does not model yet
         * (forward compatibility), and `isLenient` tolerates minor server quirks.
         */
        val CODEC: Json = Json {
            ignoreUnknownKeys = true
            isLenient = true
            explicitNulls = false
        }

        /** Builds a [JsonValue] tree from a kotlinx [JsonElement]. */
        fun fromJsonElement(element: JsonElement): JsonValue = when (element) {
            is JsonNull -> Null
            is JsonObject -> Obj(element.mapValues { fromJsonElement(it.value) })
            is JsonArray -> Arr(element.map { fromJsonElement(it) })
            is JsonPrimitive -> when {
                !element.isString && element.booleanOrNull != null -> Bool(element.booleanOrNull == true)
                !element.isString && element.doubleOrNull != null -> Num(element.doubleOrNull ?: 0.0)
                else -> Str(element.content)
            }
        }
    }
}

/**
 * Bridges [JsonValue] to kotlinx.serialization. It defers to the built-in
 * [JsonElement] serializer (both directions go through the JSON AST), so the
 * sealed type composes with every other `@Serializable` model without a hand
 * written encode/decode per case — the Kotlin equivalent of the Swift
 * `singleValueContainer` implementation.
 */
object JsonValueSerializer : KSerializer<JsonValue> {
    override val descriptor: SerialDescriptor = JsonElement.serializer().descriptor

    override fun deserialize(decoder: Decoder): JsonValue {
        val input = decoder as? JsonDecoder
            ?: throw IllegalStateException("JsonValue can only be read from JSON.")
        return JsonValue.fromJsonElement(input.decodeJsonElement())
    }

    override fun serialize(encoder: Encoder, value: JsonValue) {
        JsonElement.serializer().serialize(encoder, value.toJsonElement())
    }
}
